### ISSUE TITLE
Semantic Flow for Fast and Accurate Scene Parsing

### ISSUE THREAD

**Issue by jinfagang** [open-mmlab/mmsegmentation#196]
Semantic Flow for Fast and Accurate Scene Parsing

this model can be supported

**Comment by jinfagang** [comment]
It's the only **realtime** model achieves mIOU 80 on cityscapes.

**Comment by xvjiarui** [comment]
Hi @jinfagang 
Thanks for your suggestion. We will consider it in the next release.

**Comment by lxtGH** [comment]
@xvjiarui @jinfagang Hi thanks for your attention. I can work it for mmseg. I will make a pull request after i got the results. 
The official implementation is here: https://github.com/lxtGH/SFSegNets

**Comment by xvjiarui** [comment]
Hi @lxtGH 
How is it going?

**Comment by lxtGH** [comment]
@xvjiarui Hi ! I am training now. I will gvie the pull request if I get serveral results including Mapillary pretraining.

**Comment by Junjun2016** [comment]
> @xvjiarui Hi ! I am training now. I will gvie the pull request if I get serveral results including Mapillary pretraining.

Looking forward to it.
Could you please also have a pull request of DecoupleSegNet.

**Comment by xvjiarui** [comment]
Hi @lxtGH 
Thx! Looking forward to this PR.

**Comment by lxtGH** [comment]
I have some problems to re-produce the performance with mmseg. I try a lots of times using r18 backbone. but the results are inferior(73.x).  It's very strange.

**Comment by jinfagang** [comment]
@lxtGH  what's the *should* results of r18 and how many gap between mmseg and original one?

**Comment by lxtGH** [comment]
r18 is around 78.2 % in original repo. However, for mmseg, there are 5 % gap.  I am very confused.

**Comment by baibaidj** [comment]
> r18 is around 78.2 % in original repo. However, for mmseg, there are 5 % gap. I am very confused.

Hi. do you have any updates?

**Comment by zhangrui-wolf** [comment]
I think the main reason is that the default settings in mmseg are different from those in the paper. The biggest difference is the size of the input image, and the model in the open source code is different from that in the paper diagram. I use the same code in mmseg as in the open source repository, the highest test result is 75.05%, the average of four tests is 74.64%, and no test result is lower than 74%. Most of my super parameters are exactly the same as the default super parameters in mmseg. The input image size is 512x1024. The main difference is that I use r18-d32 backbones, and use two GPUs, each of which has four images.

**Comment by zhangrui-wolf** [comment]
I changed the parameters to train again, and found it difficult to make the result better. I think it's because they used better pre trained backbones.

**Comment by zhangrui-wolf** [comment]
The highest accuracy I got from training with the default settings of mmseg was 75.05%, and I did not train the model after that. The reasons I got from the analysis are: 1. the author used some tricks in the training, which is reflected in the code; 2. the batch size is different; 3. the parameters of the pre-trained model are different. the network structure of resnet18 is slightly different in the stem part, the stem_channel is 128. especially the third reason. I have not retrained this model since then, but I have trained other models in mmseg using the [pre-training parameters](https://drive.google.com/file/d/16mcWZSWbV3hkFWJ2cP_eJRQ6Nr1BncCp/view?usp=sharing) published by the author. By comparing the results of the same model obtained by training with mmseg's pre-training parameters, I found that the results can be improved by about 1.5% of miou.

**Comment by jinfagang** [comment]
@zhangrui-wolf I have tried original author's sf_segnets, his training result is not reproduceable. I highly doubt they correctly opensource their code or not. Even there exist some logic bug make some model can not  be trained.

They claimed it's pytorch issue, but I trained the result is not mIOU less than 78 but totally divergency, not possible to reproduce with their training strategy and loss function.

**Comment by zhangrui-wolf** [comment]
@jinfagang Instead of using his source code for training directly, I copied his core code on mmseg. 
A member of the same group was trained using the source code. 
The result will not diverge, but the training process is unstable and the accuracy can not be achieved in the end. 
If divergent, the learning rate can be adjusted according to the batch size.

**Comment by jinfagang** [comment]
@zhangrui-wolf The first epoch loss become 1.0, then after 23 epoch loss become 6.7, and then stable at 6.5....

I don't know what's model can have such weired loss cruve. At least even they converge, the mAP can not be reproducable.

However, what surprise me is the checkpoint provide by author actually get a very nice mIOU on cityscapes eval dataset... And result seems very nice.

I am so confused.

**Comment by zhangrui-wolf** [comment]
@jinfagang Have you verified his pre-training model on the test set? Is it possible that he used some trick to make the model over-fit on the verification set?

**Comment by jinfagang** [comment]
@zhangrui-wolf No, I don't eval by my hand, but the result roughly looks better than my other SOTA models. I don't know if they using some tricks, but the author himself seems not even fammilar with his code (I asked him his code has a logic bug can not be train on some model such as dsn, he claimed it's my pytorch version problem but actually same error even I using his pytorch version). 

That model is unable to reproduce original author's result. And I think original author doesn't want to fix it....

**Comment by xvjiarui** [comment]
@lxtGH Do you have any idea?

**Comment by lxtGH** [comment]
@xvjiarui Now, I got results about 77.3. However, it still 0.8 gap with original paper. I will using original backbone with stem channel 128.

**Comment by jinfagang** [comment]
@lxtGH Sounds like good

**Comment by lxtGH** [comment]
Now, it tunes to 77.9. still 0.2-0.3 gap. I will try more settings.
![image](https://user-images.githubusercontent.com/17202268/114425811-0269b000-9bec-11eb-9327-02fb517de4ed.png)

**Comment by jyang68sh** [comment]
@lxtGH Hi I'm also getting 73.x result using resnet18. I am wondering what was the change you made to get a 4 percent improve?
Thanks!
